mne_bids/utils.py

```diff
@@ -26,9 +27,144 @@
+    BIDSPath (sub-test_ses-two_task-mytask_data.csv)
+    >>> # copy and update multiple entities at once
+    >>> new_basename = bids_basename.copy().update(subject='test2', session='one')
+    >>> print(new_basename)
+    sub-test2_ses-one_task-mytask_data.csv
+    """
+
+    def __init__(self, subject=None, session=None,
+                 task=None, acquisition=None, run=None, processing=None,
+                 recording=None, space=None, prefix=None, suffix=None):
+        self.update(subject=subject, session=session, task=task,
+                    acquisition=acquisition, run=run, processing=processing,
+                    recording=recording, space=space, prefix=prefix,
+                    suffix=suffix)
+
+    @property
+    def entities(self):
+        """Return dictionary of the BIDS entities."""
+        # create an ordered dictionary of all the bids entities
+        entities = OrderedDict()
+        for key in BIDS_ENTITIES:
+            value = getattr(self, key, None)
+            if value is not None:
+                entities[key] = value
+        return entities
+
+    def __str__(self):
+        """Return the string representation of the path."""
+        basename = _gen_bids_basename(**self.entities)
+        return basename
+
+    def __repr__(self):
+        """Representation in the style of `pathlib.Path`."""
+        return "{}({!r})".format(self.__class__.__name__, str(self))
+
+    def __fspath__(self):
+        """Return the string representation for any fs functions."""
+        return str(self)
+
+    def __bytes__(self):
+        """Return the bytes representation of the path.
+
+        This is only recommended to use under Unix.
+        """
+        return os.fsencode(str(self))
+
+    def __eq__(self, other):
+        """Compare str representations."""
+        return str(self) == str(other)
+
+    def __ne__(self, other):
+        """Compare str representations."""
+        return str(self) != str(other)
+
+    def copy(self):
+        """Copy the instance.
+
+        Returns
+        -------
+        bidspath : instance of BIDSPath
+            The copied bidspath.
+        """
+        return deepcopy(self)
+
+    def update(self, **kwargs):
+        """Update inplace BIDS entity key/value pairs in object.
+
+        Parameters
+        ----------
+        kwargs :
+            Allowed BIDS entities: see `BIDS_ENTITIES`.
+
+        Returns
+        -------
+        bidspath : instance of BIDSPath
+            The bidspath.
+        """
+        # error check kwargs
+        for key, value in kwargs.items():
+            # error check allowed BIDS entity keywords
+            if key not in BIDS_ENTITIES:
+                raise ValueError('Key must be one of {BIDS_ENTITIES}, '
+                                 'got %s' % key)
+
+            # set entity value
+            setattr(self, key, value)
+
+        # run string representation to check validity of arguments
+        str(self)
```

Comment:
This seems like a bit of a hack, is there maybe a cleaner way to do it using a private function?

Reply:
I don't have a great way of integrating it off the top of my head because the `_gen_bids_basename` function does the construction of the basename and also error checking inside. 

Unless you think it should just get refactored too, but I was leaning towards the side w/o trying to do too much in this one  PR :P.